File: Data Structures & Algorithms/binary-tree-right-side-view/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        # BFS level traversal
        if not root:
            return []
        ans = []
        queue = deque([root])
        while queue:
            sz = len(queue)
            node = None
            for _ in range(sz):
                node = queue.popleft()
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            if node:
                ans.append(node.val)
        return ans

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        // find the first node where p and q are on the 2 sides
        // or the first node == p or == q
        int min = std::min(p->val, q->val);
        int max = std::max(p->val, q->val);
       TreeNode* curr = root;
       while (curr) {
        if (curr->val > max) {
            curr = curr->left;
        } else if (curr->val < min) {
            curr = curr->right;
        } else {
            return curr;
        }
       }
    }

    // TreeNode* dfs(TreeNode* root, TreeNode* p, TreeNode* q) {
    //     if (!root || root == p || root == q) {
    //         return root;
    //     }
    //     int min = std::min(p->val, q->val);
    //     int max = std::max(p->val, q->val);
    //     if (root->val > min && root->val < max) {
    //         return root;
    //     } else if (root->val > max) {
    //         return dfs(root->left, p, q);
    //     } else {
    //         return dfs(root->right, p, q);
    //     }
    // }

};
